### DISCUSSION TITLE
Simple nested progress bars

### DISCUSSION THREAD

**Discussion by nik-sm** [Textualize/rich#2272] [Q&A]
Hi - I'd love to switch from using `tqdm` for tracking progress to using `rich`.

Is there a simple recipe for tracking progress over several nested iterables?

---

I frequently have several nested loops going, and I'd like to keep their progress bars at the bottom while also printing verbose debug info above. It seems that this layout is definitely a good use case for rich.

Here's the typical code I use for `tqdm`:
```python
from tqdm import tqdm
from time import sleep

for i in tqdm(range(100), "Outer"):
    for j in tqdm(range(10), "Inner", leave=False):
        # print(f"Verbose info! {i, j}")
        sleep(0.1)
```

If no printing occurs, this works great and makes it possible to track progress of nested tasks without any development overhead. However, if you uncomment the print statement, you can see that this makes the output very ugly.

As a newbie to rich, I would have naively loved to simply do:
```python
from rich import print
from rich.progress import track
from time import sleep
for i in track(range(100), "Outer"):
    for j in track(range(10), "Inner", transient=True):
        print(f"Verbose info! {i, j}")
        sleep(0.1)
```
This or a similar recipe would totally serve my use case (tracking nested progress, code is still simple), but `rich.progress.track` currently cannot be nested.

I found this example from the docs that demonstrates having multiple bars, but I can't quite see how to adapt this to my situation without adding substantially more code, and I'm wondering if there is a simpler approach.
```python
import time

from rich.progress import Progress

with Progress() as progress:

    task1 = progress.add_task("[red]Downloading...", total=1000)
    task2 = progress.add_task("[green]Processing...", total=1000)
    task3 = progress.add_task("[cyan]Cooking...", total=1000)

    while not progress.finished:
        progress.update(task1, advance=0.5)
        progress.update(task2, advance=0.3)
        progress.update(task3, advance=0.9)
        time.sleep(0.02)
```

Thanks!

**Comment by pcroland** [comment]
```py
from rich.progress import Progress

with Progress() as pb:
    t1 = pb.add_task('inner', total=10)
    t2 = pb.add_task('outer', total=100)

    for i in range(100):
        for j in range(10):
            print(f"Verbose info! {i, j}")
            sleep(0.1)
            pb.update(task_id=t1, advance1)
        pb.update(task_id=t2, advance=1)
```
Something like this?

**Reply by nik-sm** [reply]
Hi - thanks for the reply!

Unfortunately, the "inner" bar doesn't reset after reaching 100%. 

Also - it would be ideal if the `pb.update()` calls could be avoided to keep this as simple as possible.
However, once I have a basic working example, it seems like it would be straightforward to write a wrapper that handles the incrementing, similar to `trange` or `tqdm`.

While having a look to see how `tqdm` manages the updates during iteration, I found that they actually have some experimental support for using rich to display progress bars!
(https://tqdm.github.io/docs/rich/, https://github.com/tqdm/tqdm/blob/master/tqdm/rich.py)
However, it does not support nested bars.

**Comment by pcroland** [comment]
```py
#!/usr/bin/env python3
from rich.progress import Progress
from time import sleep

with Progress() as pb:
    t1 = pb.add_task('inner', total=10)
    t2 = pb.add_task('outer', total=100)

    for i in range(100):
        for j in range(10):
            print(f"Verbose info! {i, j}")
            sleep(0.1)
            pb.update(task_id=t1, completed=j + 1)
        pb.update(task_id=t2, completed=i + 1)
```

**Reply by nik-sm** [reply]
Ok great, this works as a starting out point - thanks! I'll see if I can make a simple wrapper around this to support the same usage structure as `tqdm`.
